### PR TITLE
Transfer power level state events on room upgrade

### PR DESCRIPTION
**Pull request by anoadragon453**
Fixes https://github.com/matrix-org/synapse/issues/5580

This PR does the power levels transfer across rooms in 2 stages. The first is only copying over the user power level events (temporarily modifying that of the upgrading user to be the max power level such that they can send any events necessary).

Then, once all other state is copied over, power levels are copied from the old room again, thus reducing people back to their original power levels.

Thus, this allows for >100 PL users and state events to be copied to the new room by someone even with <100 PL (which we already allowed, it just didn't work for >100PL state event requirements).

TODO:
- [x] Add tests (https://github.com/matrix-org/sytest/pull/728)

### CODE REVIEW

**Comment by anoadragon453** [comment]
Note to self: look at `_update_upgraded_room_pls`

File: synapse/handlers/room.py [hunk 1]
```diff
@@ -271,6 +282,13 @@ def _update_upgraded_room_pls(
                 logger.warning("Unable to update PLs in old room: %s", e)
 
         logger.info("Setting correct PLs in new room")
+        if requester_original_power_level:
+            # The requester's power level was temporarily raised earlier in the upgrade
+            # process in order for them to be allowed to send the room's initial state.
+            # Return them to their original power level in the old room now
+            old_room_pl_state.content["users"][
```
Comment (erikjohnston) [inline synapse/handlers/room.py]: Won't the `old_room_state` have the correct values in it already?

File: synapse/handlers/room.py [hunk 2]
```diff
@@ -346,6 +369,35 @@ def clone_existing_room(
...
+        raised_power_level = None
+        if current_power_level < needed_power_level:
+            # Save and return this power level
+            raised_power_level = current_power_level
+
+            # Assign this power level to the requester
+            power_levels["users"][requester.user.to_string()] = needed_power_level
+
+        # Transfer the power levels via power_level_content_override instead of initial_state
+        del initial_state[(EventTypes.PowerLevels, "")]
```
Comment (erikjohnston) [inline synapse/handlers/room.py]: We probably want to take a copy of this before mutating it from under the callee?

Comment (anoadragon453) [reply synapse/handlers/room.py]: Turns out we don't even need to use the `power_level_content_override` option.

File: synapse/handlers/room.py [hunk 3]
```diff
@@ -346,6 +347,31 @@ def clone_existing_room(
             if old_event:
                 initial_state[k] = old_event.content
 
+        # Resolve the minimum power level required to send any state event
+        # We will give the upgrading user this power level temporarily (if necessary) such that
+        # they are able to copy all of the state events over, then revert them back to their
+        # original power level afterwards
```
Comment (erikjohnston) [inline synapse/handlers/room.py]: Can you just add a quick comment saying that the temporary power levels here will get overridden later?

**Pull request merged**